flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/CliTableauResultView.java

```diff
@@ -132,11 +145,23 @@ private void printResults(AtomicInteger receivedRowCount, boolean isStreamingMod
                             DEFAULT_COLUMN_WIDTH,
                             PrintUtils.NULL_COLUMN,
                             PrintUtils.ROW_KIND_COLUMN);
+            isEndOfStream = getEndOfStreamWhenRetrieveRows(change);
```

Comment:
Why you retrieve results here and continue to retrieve results in the while loop in line#186?

Reply:
Because `ResultRetrievalThread` is fetching data and put the data into the buffer by rows. Only the thread gets the mutex and the buffer is not full, the thread will put the row into the buffer.

When we try to read data from the `ChangelogCollectResult#changeRecordBuffer` in `getEndOfStreamWhenRetrieveRows`,  we may read part of the final result. Therefore, we need continue reading the data in the loop until get EOS.

Reply:
I know this, the logic you mentioned is already taken care by the codes you modified before (which only has one while loop). What I don't understand is why you split the old while loop and spread the logic of retrieving data into 3-5 places.